File: python/981_基于时间的键值存储_medium.py
class TimeMap:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.data = collections.defaultdict(list)


    def set(self, key: str, value: str, timestamp: int) -> None:
        self.data[key].append((timestamp, value))


    def get(self, key: str, timestamp: int) -> str:
        tempList = self.data[key]
        left, right = 0, len(tempList)
        while left < right:
            mid = left + (right - left) // 2
            if tempList[mid][0] <= timestamp:
                left = mid + 1
            else:
                right = mid
        if left > 0:
            return tempList[left-1][1]
        else:
            return ""



# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: cpp/981_基于时间的键值存储_medium.cpp
class TimeMap {
private:
    unordered_map <string, vector<pair<int, string>>> data;

public:
    /** Initialize your data structure here. */
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        data[key].emplace_back(timestamp, value);
    }
    
    string get(string key, int timestamp) {
        auto &keyList = data[key];
        int left = 0, right = keyList.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (keyList[mid].first <= timestamp)
                left = mid + 1;
            else
                right = mid;
        }
        if (left > 0)
            return keyList[left-1].second;
        else
            return "";

    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */